python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,280 @@
+            return
+
+        thefields = QgsFields()
+        climbindex = -1
+        descentindex = -1
+        minelevindex = -1
+        maxelevindex = -1
+        fieldnumber = 0
+
+        # Skip fields with names that are equal to the generated ones
+        for field in source.fields():
+            if field.name() == self.CLIMBATTRIBUTE:
+                feedback.pushInfo(self.tr(
+                    '{attr_name} attribute found and removed'.format(
+                        attr_name=self.CLIMBATTRIBUTE)
+                )
+                )
+                climbindex = fieldnumber
+            elif field.name() == self.DESCENTATTRIBUTE:
+                feedback.pushInfo(self.tr(
+                    '{attr_name} attribute found and removed'.format(
+                        attr_name=self.DESCENTATTRIBUTE)
+                )
+                )
+                descentindex = fieldnumber
+            elif field.name() == self.MINELEVATTRIBUTE:
+                feedback.pushInfo(self.tr(
+                    '{attr_name} attribute found and removed'.format(
+                        attr_name=self.MINELEVATTRIBUTE)
+                )
+                )
+                minelevindex = fieldnumber
+            elif field.name() == self.MAXELEVATTRIBUTE:
+                feedback.pushInfo(self.tr(
+                    '{attr_name} attribute found and removed'.format(
+                        attr_name=self.MAXELEVATTRIBUTE)
+                )
+                )
+                maxelevindex = fieldnumber
+            else:
+                thefields.append(field)
+            fieldnumber = fieldnumber + 1
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        layerwithz = source
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context, thefields,
+                                               layerwithz.wkbType(),
+                                               source.sourceCrs())
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        no_z_nodes = []
+
+        for current, feature in enumerate(features):
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            parts = feature.geometry().constParts()
```

Comment:
Check for feature.hasGeometry(), and handle appropriately for null geometry features.